Topo template fix

Yan824 opened the pull request:
Use topology template to generate topology.

Yan824 committed:
Apply patch to improve workload reading

Yan824 committed:
Improve Topology Generation with Template

Yan824 committed:
Imporve Tutorial.md

Yan824 committed:
Fix topology generation with template

Yan824 committed:
Fix topology generation with template(Improvement)

Yan824 committed:
Fix topology generation with template(Improvement)

Yan824 committed:
Fix topology generation with template(Improvement)

Yan824 committed:
Fix topology generation with template(Improvement)

Yan824 committed:
Resolve Conflicts

Yan824 committed:
Reslove Conflict with Tutorial

gabrielecastellano reviewed (changes requested):
Hi! Thank you for your contribution! I was indeed encountering some troubles with the original script for topology generation (see #81). Please refer to my review below.

### Problem 1
It is not intuitive to me that the script has all those parameters but, if using templates, they are useless (they get overridden by the template). Intuitively, one would think that it would be possible to use a template and customize it by setting parameters.

### Problem 2
You introduced a new parameter ``dualToR``. It is not clear to me the difference with the ``dual plane`` parameter. Both seem to connect a GPU to two different ASW switches (or ToRs). Is this a misunderstanding? Or an artifact due to different terminologies for different templates? In the latter case, maybe it is problematic to try to support multiple topologies with a single script.

### Problem 3
It is not clear from the tutorial the behavior of invoking the script without specifying a template. From what I have observed, it looks like a HPN-like topology is generated. However, the behavior could be different from the expected one based on the combination of ``dualToR`` and ``dual plane`` parameters. For instance, following command:
```
python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp -dualToR=1
No Template is used.
DualToR_DualPlane
64
asw_switch_num: 8
psw_switch_num: 4
Creating Topology of totally 2 segment(s), totally 1 pod(s).
AlibabaHPN_8g_2gps_DualPlane_100Gbps_A100
```
generates an HPN-like topology with dual plane (i.e., each GPU is connected to two ASW, half ASW is connected to 2 PSW, half to the other two), while the following (``dualToR`` parameter is missing):
```
$ python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp
No Template is used.
DualToR_SinglePlane
64
asw_switch_num: 8
psw_switch_num: 4
Creating Topology of totally 2 segment(s), totally 1 pod(s).
AlibabaHPN_8g_2gps_SinglePlane_100Gbps_A100
```
names the output file "SinglePlane", despite providing the parameter `--dp`, but the resulting topology is kind of a mix between single and dual plane (each GPU is connected to two ASW, but all ASW are connected to all PSW).


### Problem 4
I have tried running the example command you added in the tutorial
```
python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -topo Spectrum-X -g 8 -psn 1
```
but the result is different than the one in the figure:

```
81 8 1 72 528 H100
8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 
0 8 2880Gbps 0.000025ms 0
0 9 400Gbps 0.0005ms 0
1 8 2880Gbps 0.000025ms 0
1 10 400Gbps 0.0005ms 0
2 8 2880Gbps 0.000025ms 0
2 11 400Gbps 0.0005ms 0
3 8 2880Gbps 0.000025ms 0
3 12 400Gbps 0.0005ms 0
4 8 2880Gbps 0.000025ms 0
4 13 400Gbps 0.0005ms 0
5 8 2880Gbps 0.000025ms 0
5 14 400Gbps 0.0005ms 0
6 8 2880Gbps 0.000025ms 0
6 15 400Gbps 0.0005ms 0
7 8 2880Gbps 0.000025ms 0
7 16 400Gbps 0.0005ms 0
9 17 400Gbps 0.0005ms 0
...
9 80 400Gbps 0.0005ms 0
10 17 400Gbps 0.0005ms 0
...
16 80 400Gbps 0.0005ms 0
```

### Problem 5
The tutorial says ``for gpu number less than one segment, none psw will be used``, but this looks wrong. In the HPN topology, how can GPUs from different hosts that are attached to different ASWs could communicate without PSW? For instance, GPU 1 from host 1, is attached to ToR 1 (and ToR 2, if dual plane) and could not communicate with GPU 8 from host 136, which is attached to ToR 15 (and 16, if dual plane); to communicate, despite being in the same segment, they need Agg1 (two of them, if dual plane)
![image](https://github.com/user-attachments/assets/56839d31-6625-4249-b0a1-7bbd306479cd)

### Problem 6
I think figures are potentially ambiguous: 
- In **AlibabaHPN with Single Plane(One Pod)**, each GPU should be connected only to one ToR, while from the figure it looks like the two close dots refer to the same GPU *or nic) being connected to two ToRs (Dual Plane style).
- I would explicitly mark the two dots belonging to the same GPU in **AlibabaHPN with Dual Plane(One Pod)** just like in the HPN paper, surrounding them with a rectangle and labeling with "NIC 1" etc. (see figure above).
- Different switches have the same name but different colors in **AlibabaHPN with Single Plane(One Pod)**.

Yan824 committed:
Improvement according to pr review

Yan824 committed:
Improvement according to pr review

Yan824 committed:
Merge branch 'topo_template_fix' of https://github.com/aliyun/SimAI into topo_template_fix

Yan824 commented:
> Hi! Thank you for your contribution! I was indeed encountering some troubles with the original script for topology generation (see #81). Please refer to my review below.
> 
> ### Problem 1
> It is not intuitive to me that the script has all those parameters but, if using templates, they are useless (they get overridden by the template). Intuitively, one would think that it would be possible to use a template and customize it by setting parameters.
> 
> ### Problem 2
> You introduced a new parameter `dualToR`. It is not clear to me the difference with the `dual plane` parameter. Both seem to connect a GPU to two different ASW switches (or ToRs). Is this a misunderstanding? Or an artifact due to different terminologies for different templates? In the latter case, maybe it is problematic to try to support multiple topologies with a single script.
> 
> ### Problem 3
> It is not clear from the tutorial the behavior of invoking the script without specifying a template. From what I have observed, it looks like a HPN-like topology is generated. However, the behavior could be different from the expected one based on the combination of `dualToR` and `dual plane` parameters. For instance, following command:
> 
> ```
> python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp -dualToR=1
> No Template is used.
> DualToR_DualPlane
> 64
> asw_switch_num: 8
> psw_switch_num: 4
> Creating Topology of totally 2 segment(s), totally 1 pod(s).
> AlibabaHPN_8g_2gps_DualPlane_100Gbps_A100
> ```
> 
> generates an HPN-like topology with dual plane (i.e., each GPU is connected to two ASW, half ASW is connected to 2 PSW, half to the other two), while the following (`dualToR` parameter is missing):
> 
> ```
> $ python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp
> No Template is used.
> DualToR_SinglePlane
> 64
> asw_switch_num: 8
> psw_switch_num: 4
> Creating Topology of totally 2 segment(s), totally 1 pod(s).
> AlibabaHPN_8g_2gps_SinglePlane_100Gbps_A100
> ```
> 
> names the output file "SinglePlane", despite providing the parameter `--dp`, but the resulting topology is kind of a mix between single and dual plane (each GPU is connected to two ASW, but all ASW are connected to all PSW).
> 
> ### Problem 4
> I have tried running the example command you added in the tutorial
> 
> ```
> python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -topo Spectrum-X -g 8 -psn 1
> ```
> 
> but the result is different than the one in the figure:
> 
> ```
> 81 8 1 72 528 H100
> 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 
> 0 8 2880Gbps 0.000025ms 0
> 0 9 400Gbps 0.0005ms 0
> 1 8 2880Gbps 0.000025ms 0
> 1 10 400Gbps 0.0005ms 0
> 2 8 2880Gbps 0.000025ms 0
> 2 11 400Gbps 0.0005ms 0
> 3 8 2880Gbps 0.000025ms 0
> 3 12 400Gbps 0.0005ms 0
> 4 8 2880Gbps 0.000025ms 0
> 4 13 400Gbps 0.0005ms 0
> 5 8 2880Gbps 0.000025ms 0
> 5 14 400Gbps 0.0005ms 0
> 6 8 2880Gbps 0.000025ms 0
> 6 15 400Gbps 0.0005ms 0
> 7 8 2880Gbps 0.000025ms 0
> 7 16 400Gbps 0.0005ms 0
> 9 17 400Gbps 0.0005ms 0
> ...
> 9 80 400Gbps 0.0005ms 0
> 10 17 400Gbps 0.0005ms 0
> ...
> 16 80 400Gbps 0.0005ms 0
> ```
> 
> ### Problem 5
> The tutorial says `for gpu number less than one segment, none psw will be used`, but this looks wrong. In the HPN topology, how can GPUs from different hosts that are attached to different ASWs could communicate without PSW? For instance, GPU 1 from host 1, is attached to ToR 1 (and ToR 2, if dual plane) and could not communicate with GPU 8 from host 136, which is attached to ToR 15 (and 16, if dual plane); to communicate, despite being in the same segment, they need Agg1 (two of them, if dual plane) ![image](https://private-user-images.githubusercontent.com/7659682/404856656-56839d31-6625-4249-b0a1-7bbd306479cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0NjAxNDIsIm5iZiI6MTczNzQ1OTg0MiwicGF0aCI6Ii83NjU5NjgyLzQwNDg1NjY1Ni01NjgzOWQzMS02NjI1LTQyNDktYjBhMS03YmJkMzA2NDc5Y2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjFUMTE0NDAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTI1OGFiNjM2NmVmYzdhZDkxOGNiMmI3MGNlNmNmODJiOGUwOGNlNWUxZmM1ODc4NGQ0OGVjMTkwNTgzODIyNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UQbzdNQEBKqPW11Ra-ZLlcpo-YDBoIwbUhY4eKfTAbI)
> 
> ### Problem 6
> I think figures are potentially ambiguous:
> 
> * In **AlibabaHPN with Single Plane(One Pod)**, each GPU should be connected only to one ToR, while from the figure it looks like the two close dots refer to the same GPU *or nic) being connected to two ToRs (Dual Plane style).
> * I would explicitly mark the two dots belonging to the same GPU in **AlibabaHPN with Dual Plane(One Pod)** just like in the HPN paper, surrounding them with a rectangle and labeling with "NIC 1" etc. (see figure above).
> * Different switches have the same name but different colors in **AlibabaHPN with Single Plane(One Pod)**.

Hi! Thank you so much for your carefully review, and I have made some improvement according to them.
For problem 1, I changed the ways of parameter analysis and the users can use the template and customize them by setting parameters.
For problem 2, problem 3, I think there is some some ambiguity regarding the concepts of Dual-ToR and Dual-Plane. In the HPN paper, Dual-ToR refers to a network card's port being split into two and connected to two ToRs, with the aim of enhancing the architecture's fault recovery capability. Dual-Plane is intended to reduce hash collisions. Therefore, I added parameters related to Dual-ToR, and in the HPN template, Dual-ToR can only be enabled.

Yan824 commented:
> Hi! Thank you for your contribution! I was indeed encountering some troubles with the original script for topology generation (see #81). Please refer to my review below.
> 
> ### Problem 1
> It is not intuitive to me that the script has all those parameters but, if using templates, they are useless (they get overridden by the template). Intuitively, one would think that it would be possible to use a template and customize it by setting parameters.
> 
> ### Problem 2
> You introduced a new parameter `dualToR`. It is not clear to me the difference with the `dual plane` parameter. Both seem to connect a GPU to two different ASW switches (or ToRs). Is this a misunderstanding? Or an artifact due to different terminologies for different templates? In the latter case, maybe it is problematic to try to support multiple topologies with a single script.
> 
> ### Problem 3
> It is not clear from the tutorial the behavior of invoking the script without specifying a template. From what I have observed, it looks like a HPN-like topology is generated. However, the behavior could be different from the expected one based on the combination of `dualToR` and `dual plane` parameters. For instance, following command:
> 
> ```
> python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp -dualToR=1
> No Template is used.
> DualToR_DualPlane
> 64
> asw_switch_num: 8
> psw_switch_num: 4
> Creating Topology of totally 2 segment(s), totally 1 pod(s).
> AlibabaHPN_8g_2gps_DualPlane_100Gbps_A100
> ```
> 
> generates an HPN-like topology with dual plane (i.e., each GPU is connected to two ASW, half ASW is connected to 2 PSW, half to the other two), while the following (`dualToR` parameter is missing):
> 
> ```
> $ python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -g 8 -gt A100 -gps 2 -asn 8 -psn 4 -npa 2 --dp
> No Template is used.
> DualToR_SinglePlane
> 64
> asw_switch_num: 8
> psw_switch_num: 4
> Creating Topology of totally 2 segment(s), totally 1 pod(s).
> AlibabaHPN_8g_2gps_SinglePlane_100Gbps_A100
> ```
> 
> names the output file "SinglePlane", despite providing the parameter `--dp`, but the resulting topology is kind of a mix between single and dual plane (each GPU is connected to two ASW, but all ASW are connected to all PSW).
> 
> ### Problem 4
> I have tried running the example command you added in the tutorial
> 
> ```
> python3 ./astra-sim-alibabacloud/inputs/topo/gen_Topo_Template.py -topo Spectrum-X -g 8 -psn 1
> ```
> 
> but the result is different than the one in the figure:
> 
> ```
> 81 8 1 72 528 H100
> 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 
> 0 8 2880Gbps 0.000025ms 0
> 0 9 400Gbps 0.0005ms 0
> 1 8 2880Gbps 0.000025ms 0
> 1 10 400Gbps 0.0005ms 0
> 2 8 2880Gbps 0.000025ms 0
> 2 11 400Gbps 0.0005ms 0
> 3 8 2880Gbps 0.000025ms 0
> 3 12 400Gbps 0.0005ms 0
> 4 8 2880Gbps 0.000025ms 0
> 4 13 400Gbps 0.0005ms 0
> 5 8 2880Gbps 0.000025ms 0
> 5 14 400Gbps 0.0005ms 0
> 6 8 2880Gbps 0.000025ms 0
> 6 15 400Gbps 0.0005ms 0
> 7 8 2880Gbps 0.000025ms 0
> 7 16 400Gbps 0.0005ms 0
> 9 17 400Gbps 0.0005ms 0
> ...
> 9 80 400Gbps 0.0005ms 0
> 10 17 400Gbps 0.0005ms 0
> ...
> 16 80 400Gbps 0.0005ms 0
> ```
> 
> ### Problem 5
> The tutorial says `for gpu number less than one segment, none psw will be used`, but this looks wrong. In the HPN topology, how can GPUs from different hosts that are attached to different ASWs could communicate without PSW? For instance, GPU 1 from host 1, is attached to ToR 1 (and ToR 2, if dual plane) and could not communicate with GPU 8 from host 136, which is attached to ToR 15 (and 16, if dual plane); to communicate, despite being in the same segment, they need Agg1 (two of them, if dual plane) ![image](https://private-user-images.githubusercontent.com/7659682/404856656-56839d31-6625-4249-b0a1-7bbd306479cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0NjAxNDIsIm5iZiI6MTczNzQ1OTg0MiwicGF0aCI6Ii83NjU5NjgyLzQwNDg1NjY1Ni01NjgzOWQzMS02NjI1LTQyNDktYjBhMS03YmJkMzA2NDc5Y2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjFUMTE0NDAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTI1OGFiNjM2NmVmYzdhZDkxOGNiMmI3MGNlNmNmODJiOGUwOGNlNWUxZmM1ODc4NGQ0OGVjMTkwNTgzODIyNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UQbzdNQEBKqPW11Ra-ZLlcpo-YDBoIwbUhY4eKfTAbI)
> 
> ### Problem 6
> I think figures are potentially ambiguous:
> 
> * In **AlibabaHPN with Single Plane(One Pod)**, each GPU should be connected only to one ToR, while from the figure it looks like the two close dots refer to the same GPU *or nic) being connected to two ToRs (Dual Plane style).
> * I would explicitly mark the two dots belonging to the same GPU in **AlibabaHPN with Dual Plane(One Pod)** just like in the HPN paper, surrounding them with a rectangle and labeling with "NIC 1" etc. (see figure above).
> * Different switches have the same name but different colors in **AlibabaHPN with Single Plane(One Pod)**.

Hi! Thank you so much for your carefully review, and I have made some improvement according to them.
For problem 1, I changed the ways of parameter analysis and the users can use the template and customize them by setting parameters.https://github.com/aliyun/SimAI/commit/0d980d6769142fbdf67f6b9d34d31be4ea921a36#diff-6b1c5eceed40c2d3193924a8a80e18ee92b0e2c8c904adf8f759ff076551b2ceR525
For problem 2, problem 3, I think there is some some ambiguity regarding the concepts of Dual-ToR and Dual-Plane. In the [HPN paper](https://ennanzhai.github.io/pub/sigcomm24-hpn.pdf), Dual-ToR refers to a network card's port being split into two and connected to two ToRs, with the aim of enhancing the architecture's fault recovery capability. Dual-Plane is intended to reduce hash collisions. Therefore, I added parameters related to Dual-ToR, and in the HPN template, Dual-ToR can only be enabled.
![image](https://github.com/user-attachments/assets/448b1384-514a-45ce-b199-2b6e57a90d5a)
For Problem 4, there are discrepancy between the tutorial and code, and it is rectified. 
For Problem 5, in HPN topology, there is nvswitch inside hosts. So if GPU1 in Host 1 wants to communication with GPU8 in Host 136, it would first go to GPU 1 in Host 136, and GPU1 in Host 136 will communicate with GPU 8 in Host 136 with intra-host nvswitch. But considering that NVLink might not be enabled during SimAI simulation, I provided PSW for GPUs which numbers are less than one segment.
For Problem 6, I add the rectangle and labeling with NIC 1 for both AlibabaHPN_Dual Plane and AlibabaHPN_Single Plane, and make the index of ToRs different.
I sincerely appreciate your review once again. If you have any suggestions regarding the latest improvement, please let me know.

Yan824 committed:
Add Readme.md

Yan824 committed:
Add Readme.md

gabrielecastellano commented:
Hi! Thanks for your clarifications. Especially about ``Problem 2`` and ``Problem 3``, now it makes more sense to me.

Concerning ``Problem 5``, what you say is interesting. Checking the paper it looks like indeed hosts from different "rails" are supposed to communicate through an intermediate GPU. As you say it looks that this is not implemented in SimAI, where [all intermediate nodes in route are switches](https://github.com/aliyun/SimAI/blob/2fd25a72e3ef7e8521fd4187ad5e64ea194398e8/astra-sim-alibabacloud/astra-sim/network_frontend/ns3/common.h#L266). 

It looks like this creates a huge difference between real traffic and traffic simulated by SimAI, doesn't it? Should this be documented?

Huoyuan100861 commented:
> Hi! Thanks for your clarifications. Especially about `Problem 2` and `Problem 2`, now it makes more sense to me.
> 
> Concerning `Problem 5`, what you say is interesting. Checking the paper it looks like indeed hosts from different "rails" are supposed to communicate through an intermediate GPU. As you say it looks that this is not implemented in SimAI, where [all intermediate nodes in route are switches](https://github.com/aliyun/SimAI/blob/2fd25a72e3ef7e8521fd4187ad5e64ea194398e8/astra-sim-alibabacloud/astra-sim/network_frontend/ns3/common.h#L266).
> 
> It looks like this creates a huge difference between real traffic and traffic simulated by SimAI, doesn't it? Should this be documented?

This operation, "communicate through an intermediate GPU," is represented in the flowmodel. The specialized term for this is "PXN". If you're interested, you can search for more information about it.

gabrielecastellano commented:
> > Hi! Thanks for your clarifications. Especially about `Problem 2` and `Problem 2`, now it makes more sense to me.
> > Concerning `Problem 5`, what you say is interesting. Checking the paper it looks like indeed hosts from different "rails" are supposed to communicate through an intermediate GPU. As you say it looks that this is not implemented in SimAI, where [all intermediate nodes in route are switches](https://github.com/aliyun/SimAI/blob/2fd25a72e3ef7e8521fd4187ad5e64ea194398e8/astra-sim-alibabacloud/astra-sim/network_frontend/ns3/common.h#L266).
> > It looks like this creates a huge difference between real traffic and traffic simulated by SimAI, doesn't it? Should this be documented?
> 
> This operation, "communicate through an intermediate GPU," is represented in the flowmodel. The specialized term for this is "PXN". If you're interested, you can search for more information about it.

I have seen there is an env parameter ``AS_PXN_ENABLE``. Could you explain the behavior? For now, I have run simulations without this variable set. From the fct output file I can see all communications that happen within a server (due to TP) and from GPUs of the same rail (due to DP), but I don't see indeed transfers between GPUs that have both different rails and servers.
Is this because the communication was already split into two according with the behavior described in HPN? What does ``AS_PXN_ENABLE`` would change?

Thank you for the clarification

Huoyuan100861 commented:
> > > Hi! Thanks for your clarifications. Especially about `Problem 2` and `Problem 2`, now it makes more sense to me.
> > > Concerning `Problem 5`, what you say is interesting. Checking the paper it looks like indeed hosts from different "rails" are supposed to communicate through an intermediate GPU. As you say it looks that this is not implemented in SimAI, where [all intermediate nodes in route are switches](https://github.com/aliyun/SimAI/blob/2fd25a72e3ef7e8521fd4187ad5e64ea194398e8/astra-sim-alibabacloud/astra-sim/network_frontend/ns3/common.h#L266).
> > > It looks like this creates a huge difference between real traffic and traffic simulated by SimAI, doesn't it? Should this be documented?
> > 
> > 
> > This operation, "communicate through an intermediate GPU," is represented in the flowmodel. The specialized term for this is "PXN". If you're interested, you can search for more information about it.
> 
> I have seen there is an env parameter `AS_PXN_ENABLE`. Could you explain the behavior? For now, I have run simulations without this variable set. From the fct output file I can see all communications that happen within a server (due to TP) and from GPUs of the same rail (due to DP), but I don't see indeed transfers between GPUs that have both different rails and servers. Is this because the communication was already split into two according with the behavior described in HPN? What does `AS_PXN_ENABLE` would change?
> 
> Thank you for the clarification

Thanks for your attention and support to SimAI. We welcome you to join our community group (there's a QR code at the end of the README on the homepage) for communication and discussion.

gabrielecastellano commented:
> > > > Hi! Thanks for your clarifications. Especially about `Problem 2` and `Problem 2`, now it makes more sense to me.
> > > > Concerning `Problem 5`, what you say is interesting. Checking the paper it looks like indeed hosts from different "rails" are supposed to communicate through an intermediate GPU. As you say it looks that this is not implemented in SimAI, where [all intermediate nodes in route are switches](https://github.com/aliyun/SimAI/blob/2fd25a72e3ef7e8521fd4187ad5e64ea194398e8/astra-sim-alibabacloud/astra-sim/network_frontend/ns3/common.h#L266).
> > > > It looks like this creates a huge difference between real traffic and traffic simulated by SimAI, doesn't it? Should this be documented?
> > > 
> > > 
> > > This operation, "communicate through an intermediate GPU," is represented in the flowmodel. The specialized term for this is "PXN". If you're interested, you can search for more information about it.
> > 
> > 
> > I have seen there is an env parameter `AS_PXN_ENABLE`. Could you explain the behavior? For now, I have run simulations without this variable set. From the fct output file I can see all communications that happen within a server (due to TP) and from GPUs of the same rail (due to DP), but I don't see indeed transfers between GPUs that have both different rails and servers. Is this because the communication was already split into two according with the behavior described in HPN? What does `AS_PXN_ENABLE` would change?
> > Thank you for the clarification
> 
> Thanks for your attention and support to SimAI. We welcome you to join our community group (there's a QR code at the end of the README on the homepage) for communication and discussion.

Thanks. I will check that and join the discussion. However, it would be great to have some info directly available here on the Repo for everyone's convenience.

BTW I had more time to review the new topo gen script and noticed a bug concerning DCN+ topologies; please check my PR #85 that fixes that.
